graphql_compiler/cost_estimation/statistics.py

```diff
@@ -0,0 +1,211 @@
+        """Return number of edges of given class connecting two vertex types.
+
+        For cost estimation, sometimes we need to predict traversal costs between two vertex types
+        using an edge type that connects the vertices' abstract types. If this statistic isn't
+        provided, it can be roughly estimated using multiple calls of get_class_count().
+
+        Args:
+            vertex_out_class_name: str, vertex class name.
+            vertex_in_class_name: str, vertex class name.
+            edge_class_name: str, edge class name.
+
+        Returns:
+            - int, the count of edges if the statistic exists
+            - None otherwise
+        """
+        return None
+
+    @abstractmethod
+    def get_domain_count(self, vertex_name, field_name):
+        """Return the number of distinct values the vertex field has over all vertex instances.
+
+        This statistic helps estimate the result size of @filter directives used to restrict values
+        using equality operators like '=', '!=', and 'in_collection'.
+
+        Preconditions:
+        1. field_name must be a valid vertex field of vertex_name
+
+        Args:
+            vertex_name: str, name of a vertex.
+            field_name: str, name of a vertex field.
+
+        Returns:
+            - int, domain size of the vertex field's values if the statistic exists
+            - None otherwise
+        """
+        return None
+
+    @abstractmethod
+    def get_histogram(self, vertex_name, field_name):
+        """Return a histogram for the given vertex field, providing statistics about range values.
+
+        This statistic helps estimate the result size of @filter directives used to restrict
+        numbers using inequality operators like '>', '<', and 'between'.
+
+        Preconditions:
+        1. field_name must be a valid vertex field of vertex_name
+
+        Args:
+            vertex_name: str, name of a vertex.
+            field_name: str, name of a vertex field.
+
+        Returns:
+            - list[tuple(int, int, int)], histogram as a list of tuples with 3 elements each if the
+              statistic exists. Each tuple represents a bin as (start, end, elementCount), where
+              [start, end) is the range of values in the bin, and elementCount is the bin element
+              count.
+            - None otherwise
+        """
+        return None
+
+
+class LocalStatistics(Statistics):
+    """Provides statistics using ones given at initialization."""
+    def __init__(self, class_count, edge_count_between_vertex_pair, domain_count, histogram):
+        """Initializes statistics with the given data.
+
+        Args:
+            class_count: dict, str -> int, mapping vertex/edge class name to class count.
+            edge_count_between_vertex_pair: dict, (str, str, str) -> int,
+                mapping tuple of (vertex out class name, vertex in class name, edge class name) to
+                count of edge type between vertex types.
```

Comment:
nit: count of edges between the two vertices. You are not really referring to types here but concrete values.

Reply:
The types are not necessarily concrete, for example, we might have this statistic between abstract to abstract vertices. Even though we won't use the abstract to abstract statistics in the short term, I don't see why we should artificially limit ourselves.